openedx/course-discovery
# feat: csv data loader

### [PROD-2613](https://openedx.atlassian.net/browse/PROD-2613)

### Description

This PR adds a management command that uses CSVDataLoader, also added via this PR, to import course and course information from CSV into discovery. For a single row, the CSV loader works as follows:

- Check the existence of the course and course run in the database. They should be draft if already present
- If the course and course run are not present, create an initial entry
- Update course and course run through separate API calls with their respective fields. 
- Mark the run as reviewed. This essentially publishes the course if applicable.

The loader uses existing course and course run view sets to make API calls to create or update objects. The consumption of APIs as opposed to writing ORM queries is that the CSV loaded is mimicking the way courses and course runs are authored on the publisher. Eventually, all the course/course runs will be accessible and editable from the publisher, so the data representation of API will help achieve that. Also, the APIs already have certain data validation and status update procedures in place which will ensure no bad/missing information is being ingested into discovery.

### Assumptions

The csv loader works under some assumptions, at least at the time of creating this PR:
- Course and course run should be draft if already present before ingesting.
- LevelType and related subjects should be present in the database
- Collaborators(for the course) and Staff(for course runs) are created during ingestion if not located within DB. Staff's name is saved only
- The dates are in UTC as compared to EST. This might require a change in the near future if the provided dates in the sample are in EST
- If the content and/or transcript language are not in IETF language format, course run data is not updated
- Some CSV keys might need an update once there is a sample data set
- The default course editor of the imported course will be the user associated with OAuth application used for making API calls.
 

### Testing

You can use the management command to check the loader, replace csv path value with a valid csv path.
```
python manage.py import_course_metadata --csv_path='/edx/app/discovery/discovery/course_discovery/apps/course_metadata/data_loaders/test.csv'
```

Make sure you have the following LevelType entries in the database:
- Introductory
- Advanced
- Intermediate

Also, add the following subjects to the DB, if not present:

- Mathematics
- Physics
- Computer Science
- Social Sciences

### Test CSVs
- [test.csv](https://github.com/openedx/course-discovery/files/7904309/test.csv)
- [test.csv](https://github.com/openedx/course-discovery/files/7904316/test.csv)

### Testing on Sandbox

dawoudsheraz.sandbox.edx.org/

- Download the CSV generator from ticket https://openedx.atlassian.net/browse/PROD-2613?focusedCommentId=589805
- Generate a new CSV
- Connect to VPN
- Copy the csv to sandbox via command `scp test.csv <github_user>@dawoudsheraz.sandbox.edx.org:/tmp`(Assuming command is run from directory where csv is present)
- SSh into sandbox `ssh <github_user>@dawoudsheraz.sandbox.edx.org:`. 
- cd `/edx/app/discovery/discovery`
- run `source ../venvs/discovery/bin/activate`
- Run `python manage.py import_course_metadata --csv_path='/tmp/test.csv'`
- Verify the data on https://discovery-dawoudsheraz.sandbox.edx.org/admin (email: admin@example.com, edx)
- Verify the data on LMS/Studio (admin@example.com, Hamilton)
- Verify data on ecommerce

### Todos
to be catered in follow-up PRs:

- Change price to proper enrollment track-type instead of hard-coding to verified
- Update how the staff/person ids are passed
- Update date parsing to EST once the time values are confirmed
- Add verified deadline override verification checks

Also, in the ticket, there is a script to generate CSV with dummy data. You should generate some other test csv using that script and verify the behavior.
